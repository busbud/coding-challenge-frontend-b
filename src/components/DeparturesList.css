.departures-list {
  list-style-type: none;
  padding: 0;
}
.departures-list-item {
  margin-bottom: 2rem;
  font-weight: 300;
  animation: pop-in 0.5s;
}
.departures-list-item:nth-child(even) {
  animation: pop-in-inverse 0.5s;
}
.departures-list-item__row {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}
.departures-list-item__departure-time {
  color: #199ceb;
}
.departures-list-item__arrival-time {
  color: rgba(10, 59, 95, 0.54);
}

@keyframes pop-in {
  0% {
    opacity: 0;
    transform: translateX(-500px);
  }
  100% {
    opacity: 1;
    transform: translateX(0px);
  }
}

@keyframes pop-in-inverse {
  0% {
    opacity: 0;
    transform: translateX(500px);
  }
  100% {
    opacity: 1;
    transform: translateX(0px);
  }
}
