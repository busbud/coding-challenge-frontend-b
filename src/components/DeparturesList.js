import React from 'react';
import PropTypes from 'prop-types';
import { map } from 'ramda';
import Departure from './Departure';

const DeparturesList = ({ departures }) => {
  return (
    <div className="App-body">
      {map(departureInfo => <Departure key={departureInfo.id} {...departureInfo} />, departures)}
    </div>
  );
};

DeparturesList.propTypes = {
  departures: PropTypes.arrayOf(PropTypes.object),
};

DeparturesList.defaultProps = {
  departures: [],
};

export default DeparturesList;
