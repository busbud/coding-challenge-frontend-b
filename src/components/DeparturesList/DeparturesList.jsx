import React, { useContext, useRef } from "react";
import { useTranslation } from "react-i18next";

import { AppContext } from "context";
import { currencyFormatter } from "utils/price";
import { formattedDate } from "helpers/date";
import Container from "commons/Container";
import FakeLoader from "commons/FakeLoader";

import {
  DeparturesWrapper,
  OperatorImage,
  Box,
  CityBox,
  CityBoxContent,
  CityBoxTitle,
  CityBoxName,
  CityBoxTime,
  CityBoxOverlay,
  OperatorBox,
  OperatorPrice,
} from "./styles";

const DeparturesList = () => {
  const { t } = useTranslation();
  const myRef = useRef(null);
  const {
    departures = [],
    locations = [],
    operators = [],
    cities = [],
    loading = false,
    searched = false,
  } = useContext(AppContext);

  if (!searched) {
    return <></>;
  }

  if (searched && loading) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresSearchMessage")}</h2>
          <Box>
            <CityBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </CityBox>
            <CityBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </CityBox>
            <OperatorBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </OperatorBox>
          </Box>
        </Container>
      </DeparturesWrapper>
    );
  }

  if (
    searched &&
    !loading &&
    !departures &&
    !locations &&
    !operators &&
    !cities
  ) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresNotFoundMessage")}</h2>
        </Container>
      </DeparturesWrapper>
    );
  }

  if (
    !departures?.length ||
    !locations?.length ||
    !operators?.length ||
    !cities?.length
  ) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresErrorMessage")}</h2>
        </Container>
      </DeparturesWrapper>
    );
  }

  const reducedDepartures = departures.reduce((departuresArray, departure) => {
    const destination = locations.find(
      ({ id }) => departure.destination_location_id === id
    );
    const origin = locations.find(
      ({ id }) => departure.origin_location_id === id
    );
    const operator = operators.find(({ id }) => departure.operator_id === id);

    origin.city = cities.find(({ id }) => origin.city_id === id);
    destination.city = cities.find(({ id }) => destination.city_id === id);

    departuresArray.push({
      ...departure,
      origin,
      destination,
      operator,
    });

    return departuresArray;
  }, []);

  myRef?.current?.scrollIntoView();

  return (
    <DeparturesWrapper ref={myRef}>
      <Container>
        <h2>{t("departuresFound")}</h2>
        {reducedDepartures.map(
          ({
            id,
            operator,
            origin,
            departure_time,
            destination,
            arrival_time,
            prices,
          }) => (
            <Box key={id}>
              <CityBox image_url={origin.city.hero_image_url}>
                <CityBoxOverlay />
                <CityBoxContent>
                  <CityBoxTitle>{t("origin")}</CityBoxTitle>
                  <div>
                    <CityBoxTime>{formattedDate(departure_time)}</CityBoxTime>
                    <CityBoxName>
                      {origin.name} - {origin.city.name}
                    </CityBoxName>
                  </div>
                </CityBoxContent>
              </CityBox>
              <CityBox image_url={destination.city.hero_image_url}>
                <CityBoxOverlay />
                <CityBoxContent>
                  <CityBoxTitle>{t("destination")}</CityBoxTitle>
                  <div>
                    <CityBoxTime>{formattedDate(arrival_time)}</CityBoxTime>
                    <CityBoxName>
                      {destination.name} - {destination.city.name}
                    </CityBoxName>
                  </div>
                </CityBoxContent>
              </CityBox>
              <OperatorBox>
                <OperatorImage src={operator.logo_url} alt={operator.name} />
                <OperatorPrice>{currencyFormatter(prices.total)}</OperatorPrice>
              </OperatorBox>
            </Box>
          )
        )}
      </Container>
    </DeparturesWrapper>
  );
};

export default DeparturesList;
