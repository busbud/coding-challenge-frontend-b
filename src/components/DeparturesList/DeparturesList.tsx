import React from 'react';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import styled from 'styled-components';
import Container from '@material-ui/core/Container';
import { FormattedNumber, FormattedTime } from 'react-intl';
import {
  AugmentedDeparture,
  DepartureInformation
} from '../../store/departures';

interface Props {
  departures: AugmentedDeparture[];
  information: DepartureInformation;
}

const StyledPaper = styled(Paper)`
  display: flex;
  justify-content: space-between;
  padding: ${props => props.theme.busbud.card.padding};
`;

const StyledTime = styled.time`
  display: inline-block;
  min-width: 70px;
`;

const StyledArrow = styled.div`
  font-weight: bold;
  margin-bottom: 4px;
`;

const Location = styled.div`
  @media (min-width: 768px) {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
`;

const Price = styled.div`
  border-left: 1px solid ${props => props.theme.busbud.color.grey};
  font-size: 1.1rem;
  margin-left: 10px;
  padding-left: 10px;
`;

export default function Departures({ departures = [], information }: Props) {
  return (
    <Container>
      <Grid container spacing={3}>
        {departures.map(departure => {
          const {
            id,
            prices: { total, currency },
            departure_time: departureTime,
            arrival_time: arrivalTime,
            originLocationName,
            destinationLocationName
          } = departure;
          return (
            <Grid item xs={12} md={6} key={id}>
              <StyledPaper>
                <div>
                  <Location>
                    <strong>
                      <StyledTime>
                        <FormattedTime value={departureTime} />
                      </StyledTime>{' '}
                      {information.originCity.name}
                    </strong>{' '}
                    - {originLocationName}
                  </Location>
                  <StyledArrow>&darr;</StyledArrow>
                  <Location>
                    <strong>
                      <StyledTime>
                        <FormattedTime value={arrivalTime} />
                      </StyledTime>{' '}
                      {information.destinationCity.name}
                    </strong>{' '}
                    - {destinationLocationName}
                  </Location>
                </div>
                <Price>
                  <FormattedNumber
                    // eslint-disable-next-line react/style-prop-object
                    style="currency"
                    value={total / 100}
                    currency={currency}
                  />
                </Price>
              </StyledPaper>
            </Grid>
          );
        })}
      </Grid>
    </Container>
  );
}
