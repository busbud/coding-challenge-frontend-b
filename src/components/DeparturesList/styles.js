import styled from "styled-components";

import { BREAKPOINTS } from "utils/constants";

export const DeparturesWrapper = styled.div`
  padding-bottom: 48px;
`;

export const OperatorImage = styled.img`
  width: 150px;
  display: block;
  height: auto;
`;

export const Box = styled.div`
  width: 100%;
  border-radius: 4px;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
  overflow: hidden;

  &:not(:last-of-type) {
    margin-bottom: 32px;
  }

  @media screen and (min-width: ${BREAKPOINTS.mobile}) {
    align-items: center;
    display: flex;
  }

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    border-radius: none;
    box-shadow: none;
    overflow: initial;

    & > div:not(:last-of-type) {
      margin-right: 40px;
    }
  }
`;

export const CityBox = styled.div`
  height: 150px;
  align-items: center;
  background-color: var(--white);
  background-image: url(${({ image_url }) => `${image_url}`});
  background-position: center;
  background-size: cover;
  flex: 1;
  position: relative;
  padding: ${({ loading }) => (loading ? "8px" : "0px")};

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    border-radius: 4px;
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);

    &:after {
      width: 24px;
      border-top: 4px var(--black) dotted;
      content: "";
      display: block;
      position: absolute;
      opacity: 0.2;
      top: 50%;
      right: -32px;
    }
  }
`;

export const CityBoxOverlay = styled.div`
  width: 100%;
  height: 100%;
  background-color: var(--black);
  border-radius: 4px;
  display: block;
  position: absolute;
  opacity: 0.3;
  top: 0;
  left: 0;
  z-index: 1;

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    border-radius: 4px;
  }
`;

export const CityBoxContent = styled.div`
  height: 100%;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  position: relative;
  z-index: 2;
  padding: 16px;
`;

export const CityBoxTitle = styled.h3`
  color: var(--white);
  font-size: 16px;
  font-weight: 700;
`;

export const CityBoxTime = styled.span`
  color: var(--white);
  display: block;
  font-size: 18px;
  font-weight: 900;
`;

export const CityBoxName = styled.span`
  background-color: var(--white);
  color: var(--black);
  font-size: 14px;
  font-weight: 900;
  padding: 2px 4px;
`;

export const OperatorBox = styled.div`
  height: 150px;
  align-items: center;
  background-color: var(--white);
  display: ${({ loading }) => (loading ? "block" : "flex")};
  justify-content: space-evenly;
  flex-direction: column;
  flex: 1;
  padding: ${({ loading }) => (loading ? "8px" : "16px")};

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    border-radius: 4px;
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
  }
`;

export const OperatorPrice = styled.span`
  color: var(--blue);
  font-size: 18px;
  font-weight: 700;
`;
