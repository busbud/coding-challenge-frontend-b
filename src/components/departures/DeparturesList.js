import React from 'react'
import _ from 'underscore'

class DeparturesList extends React.Component {
  constructor(props){
    super(props)

    this.state = props
  }

  componentWillReceiveProps(nextProps){
    console.log('nextProps')
    console.log(nextProps)

    if(! (_.isEqual(this.state, nextProps))){
      this.setState({ departures: nextProps.departures })
    }
  }

  render(){
    const { departures } = this.state
    console.log(departures)

    return(
      <div>
        {
          _.map(departures, function(value){
            return(
              <h1>{value.arrival_timezone}</h1>
            )
          })
        }
      </div>
    )
  }
}

export default DeparturesList
