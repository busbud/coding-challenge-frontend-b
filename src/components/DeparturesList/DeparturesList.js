import React, { PropTypes } from 'react';
import DepartureContainer from '../Departure/DepartureContainer';
import Loader from '../Loader/Loader';
import { orderBy }from 'lodash';

import './DeparturesList.scss';

class DeparturesList extends React.Component {
  constructor(props) {
    super(props);
  }
  getSortedDepartures(departures, sort) {
    //return an empty array if there are no departures
    if (!departures || !departures.length) {
      return [];
    }

    try {
      //sort depending on the sort type
      switch (sort) {
        case 'departureDate': 
          //order by dateDeparture ASC
          return orderBy(departures, 'display.departure_time');
        case 'price': 
          //order by price ASC
          return orderBy(departures, 'display.price');
        case 'company': 
          //order by operator's name ASC
          return orderBy(departures, 'display.operator.display_name');
        default:
        return departures;
      }
    } catch(err) {
      //in case of error (if we were not able to get all necessary parameters, just return the original array)
      return departures;
    }

  }
  render() {
    const { isFetching, data, sort } = this.props;

    //get sorted departures
    let departures = this.getSortedDepartures(data.departures, sort);

    return (
      <div className="departures-list">
        {departures.map((departure) => {
            return <DepartureContainer key={departure.id} departure={departure.display} />;
        })}
        {isFetching ? <Loader /> : null }
      </div>
    );
  }
}

DeparturesList.propTypes = {
    isFetching: PropTypes.bool.isRequired,
    data: PropTypes.object.isRequired,
    sort: PropTypes.string.isRequired
};

export default DeparturesList;