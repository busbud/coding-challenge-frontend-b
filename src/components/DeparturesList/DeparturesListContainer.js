import { connect } from 'react-redux';
import DeparturesList from './DeparturesList';

/**
 *  Container for DeparturesList : map props to redux store
 **/

const mapStateToProps = (state) => {
  return {
    isFetching: state.api.isFetching,
    data: state.api.data,
    sort: state.sort
  };
};

const mapDispatchToProps = (dispatch) => {
  return {};
};

const DeparturesListContainer = connect(mapStateToProps, mapDispatchToProps)(DeparturesList);

export default DeparturesListContainer;