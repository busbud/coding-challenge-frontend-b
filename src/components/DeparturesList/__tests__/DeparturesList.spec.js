/* eslint-disable jest/no-mocks-import */
import React from "react";
import { render } from "@testing-library/react";

import "i18n";

import { AppContext } from "context";
import { MOCKED_CONTEXT } from "context/__mocks__";

import DeparturesList from "../";

describe("<DeparturesList />", () => {
  it("should render properly", () => {
    const { container } = render(
      <AppContext.Provider value={MOCKED_CONTEXT}>
        <DeparturesList />
      </AppContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render a load structure when the search is loading", () => {
    const contextValue = {
      departures: [],
      locations: [],
      operators: [],
      cities: [],
      loading: true,
      searched: true,
    };

    const { getByText } = render(
      <AppContext.Provider value={contextValue}>
        <DeparturesList />
      </AppContext.Provider>
    );

    expect(getByText("Searching for departures...")).toBeInTheDocument();
  });

  it("should render a not found message if the search return nothing", () => {
    const contextValue = {
      departures: null,
      locations: null,
      operators: null,
      cities: null,
      loading: false,
      searched: true,
    };

    const { getByText } = render(
      <AppContext.Provider value={contextValue}>
        <DeparturesList />
      </AppContext.Provider>
    );

    expect(
      getByText("Sorry, we didn't found departures for your search")
    ).toBeInTheDocument();
  });

  it("should render a error message if some of the arrays are missing", () => {
    const contextValue = {
      departures: [{ id: 1 }],
      locations: [{ id: 1 }],
      operators: [{ id: 1 }],
      cities: null,
      searched: true,
    };

    const { getByText } = render(
      <AppContext.Provider value={contextValue}>
        <DeparturesList />
      </AppContext.Provider>
    );

    expect(
      getByText("Something went wrong! Please try your search again.")
    ).toBeInTheDocument();
  });
});
