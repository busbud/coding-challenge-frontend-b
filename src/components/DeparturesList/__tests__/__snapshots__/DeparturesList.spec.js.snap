// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeparturesList /> should render properly 1`] = `
<div>
  <div
    class="sc-dlfnbm gcgkll"
  >
    <div
      class="sc-bdfBwQ kZkkxR"
    >
      <h2>
        Departures found
      </h2>
      <div
        class="sc-eCssSg cNuKOQ"
      >
        <div
          class="sc-jSgupP gwIvAm"
        >
          <div
            class="sc-gKsewC hVeZal"
          />
          <div
            class="sc-iBPRYJ kaGFBF"
          >
            <h3
              class="sc-fubCfw dcLfJZ"
            >
              Origin
            </h3>
            <div>
              <span
                class="sc-pFZIQ eQauFK"
              >
                Sat, January 30 - 9:30 am
              </span>
              <span
                class="sc-jrAGrp gzhVfY"
              >
                Gare du Palais
                 - 
                Quebec City
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-jSgupP dkBRuJ"
        >
          <div
            class="sc-gKsewC hVeZal"
          />
          <div
            class="sc-iBPRYJ kaGFBF"
          >
            <h3
              class="sc-fubCfw dcLfJZ"
            >
              Destination
            </h3>
            <div>
              <span
                class="sc-pFZIQ eQauFK"
              >
                Sat, January 30 - 12:45 pm
              </span>
              <span
                class="sc-jrAGrp gzhVfY"
              >
                Gare d'autocars
                 - 
                Montreal
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-kEjbxe cOSuxd"
        >
          <img
            alt="OrlÃ©ans Express"
            class="sc-hKgILt fZkQDV"
            src="https://busbud.imgix.net/operator-logos/logo_orleans-express.png.png?h={height}&w={width}&auto=format&fit=fill&bg=0FFF"
          />
          <span
            class="sc-iqHYGH frYZlH"
          >
            $5,549.00
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
